doc/userguide/capture-hardware/af-xdp.rst

```diff
@@ -0,0 +1,279 @@
+to the given NUMA node.
+
+Memory assigned per socket/thread is 16MB, so each worker thread requires at least
+16MB of free space. As stated above hugepages can be of various sizes, consult the
+OS to confirm with ``cat /proc/meminfo``.
+
+Example ::
+  
+    8 worker threads * 16Mb = 128Mb
+    hugepages = 2048 kB
+    so: pages required = 62.5 (63) pages
+
+See https://www.kernel.org/doc/Documentation/vm/hugetlbpage.txt for detailed
+description.
+
+To enable unaligned chunk mode:
+
+::
+
+  af-xdp:
+    mem-unaligned: <yes/no>
+    mem-unaligned: yes
+
+enable-busy-poll
+~~~~~~~~~~~~~~~~
+
+Enables or disables busy polling.
+
+::
+
+  af-xdp:
+    enable-busy-poll: <yes/no>
+    enable-busy-poll: yes
+
+busy-poll-time
+~~~~~~~~~~~~~~
+
+Sets the approximate time in microseconds to busy poll on a ``blocking receive``
+when there is no data.
+
+::
+
+  af-xdp:
+    busy-poll-time: <time>
+    busy-poll-time: 20
+
+busy-poll-budget
+~~~~~~~~~~~~~~~~
+
+Budget allowed for batching of ingress frames. Larger values means more
+frames can be stored/read. It is recommended to test this for performance.
+
+::
+
+  af-xdp:
+    busy-poll-budget: <budget>
+    busy-poll-budget: 64
+
+Linux tunables
+~~~~~~~~~~~~~~~
+
+The ``SO_PREFER_BUSY_POLL`` option works in concert with the following two Linux
+knobs to ensure best capture performance. These are not socket options:
+
+- gro-flush-timeout
+- napi-defer-hard-irq
+
+The purpose of these two knobs is to defer interrupts and to allow the
+NAPI context to be scheduled from a watchdog timer instead
+``(gro-flush-timeout)``. The ``napi-defer-hard-irq`` indicates the number of
+attempts before exiting. When enabled the softirq NAPI context will
+exit early and allow busy polling.
+
+When no traffic is received for ``gro-flush-timeout`` this watchdog will
+timeout and softirq handling will resume.
+
+::
+
+  af-xdp:
+    gro-flush-timeout: 2000000
+    napi-defer-hard-irq: 2
+
+
+Hardware setup
+---------------
+
+Intel NIC setup
+~~~~~~~~~~~~~~~
+
+Intel network cards don't support symmetric hashing but it is possible to emulate
+it by using a specific hashing function.
+
+Follow these instructions closely for desired result::
+
+ ifconfig eth3 down
+
+Enable symmetric hashing ::
+
+ ifconfig eth3 down 
+ ethtool -L eth3 combined 16 # if you have at least 16 cores
+ ethtool -K eth3 rxhash on 
+ ethtool -K eth3 ntuple on
+ ifconfig eth3 up
+ ./set_irq_affinity 0-15 eth3
+ ethtool -X eth3 hkey 6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A:6D:5A equal 16
+ ethtool -x eth3
+ ethtool -n eth3
+
+In the above setup you are free to use any recent ``set_irq_affinity`` script. It is available in any Intel x520/710 NIC sources driver download.
+
+**NOTE:**
+We use a special low entropy key for the symmetric hashing. `More info about the research for symmetric hashing set up <http://www.ndsl.kaist.edu/~kyoungsoo/papers/TR-symRSS.pdf>`_
```

Comment:
Not sure what "low entropy" means in this context? The reason for the hash key is to get packets from both sides of the flow on to the same CPU which is a requirement for suri's flow-tracking

Reply:
I mean, I get that it's structured and predictable, but that's the whole point and seems like un-necessarily dense jargon :)

Reply:
I copied this section from ``doc/userguide/capture-hardware/ebpf-xdp.rst`` as I wanted the documentation to be consistent.